cpp/src/arrow/engine/substrait/serde.h

```diff
@@ -0,0 +1,79 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This API is EXPERIMENTAL.
+
+#pragma once
+
+#include <functional>
+#include <string>
+#include <vector>
+
+#include "arrow/buffer.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/options.h"
+#include "arrow/engine/substrait/extension_set.h"
+#include "arrow/engine/visibility.h"
+#include "arrow/result.h"
+#include "arrow/util/string_view.h"
+
+namespace arrow {
+namespace engine {
+
+using ConsumerFactory = std::function<std::shared_ptr<compute::SinkNodeConsumer>()>;
```

Comment:
I mentioned this elsewhere but now I understand it better.  I think we could name this `RelationConsumer` or something like that.  Or add some documentation.  It isn't clear at a first glance that Arrow ExecPlan's "sink" maps to a "relation".

That being said, I don't know if this feels quite right.  Shouldn't a substrait plan include a "sink" relation for handling the data?  For example, the relation could be a "write" relation or it could be a "send out via flight" relation or it could be a "send IPC to some named Arrow C bridge".

However, I suppose those would all be custom relations at the moment (Substrait really hasn't defined that part) so I think what we have here is fine.

Reply:
Do you think the comments I added earlier in 9453707 are good enough to resolve this?

My intuition for why Substrait is like this is that it models the query part of a query-response interactions, where the response is the resulting table (or, I guess, one or mote tables in this case). Note however that they do have a [write](https://substrait.io/relations/logical_relations/#write-operator) relation that does more or less what you're referring to (though it's not nearly flexible enough to represent those things yet), so I think they have thought of this already. With respect to custom relations, there is already something for it in the [protobuf files](https://github.com/substrait-io/substrait/blob/5bf92b965d9c5f0e73564e9b7e6cfed34399a580/proto/substrait/relations.proto#L203-L221).

Reply:
I've come to terms with it.  I think I need to understand / internalize more what it means for a Substrait plan to have multiple root relations.  Let's not worry about it now.  The comments you added are very helpful.